chia/data_layer/data_layer.py

```diff
@@ -224,105 +183,91 @@ async def fetch_and_validate(self, subscription: Subscription) -> None:
-            is_valid = False
-
-        # If for some reason we have mismatched data using the light checks, recheck all history as a fallback.
-        if not is_valid:
-            self.log.warning(f"Light validation failed for {tree_id}. Validating all history.")
-            to_check = await self.wallet_rpc.dl_history(launcher_id=tree_id, min_generation=uint32(1))
             try:
-                is_valid = await self._validate_batch(tree_id, to_check, 0)
+                success = await insert_from_delta_file(
+                    self.data_store,
+                    subscription.tree_id,
+                    root.generation,
+                    [record.root for record in reversed(to_download)],
+                    ip,
+                    port,
+                    self.server_files_location,
+                    self.log,
+                )
+                if success:
+                    self.log.info(
+                        f"Finished downloading and validating {subscription.tree_id}. "
+                        f"Wallet generation saved: {singleton_record.generation}. "
+                        f"Root hash saved: {singleton_record.root}."
+                    )
+                    break
             except asyncio.CancelledError:
                 raise
+            except aiohttp.client_exceptions.ClientConnectorError:
+                self.log.warning(f"Server {ip}:{port} unavailable for {tree_id}.")
             except Exception as e:
-                self.log.error(f"Error in validate batch for {tree_id}: {e}")
-                is_valid = False
-            if not is_valid:
-                await self.data_store.set_validated_wallet_generation(tree_id, 0)
-                await self.data_store.rollback_to_generation(tree_id, 0)
-                raise RuntimeError("Could not validate on-chain data. Downloading from scratch as a fallback.")
-
-        self.log.info(
-            f"Finished downloading and validating {subscription.tree_id}. "
-            f"Wallet generation saved: {singleton_record.generation}. "
-            f"Root hash saved: {singleton_record.root}."
-        )
-        await self.data_store.set_validated_wallet_generation(tree_id, int(singleton_record.generation))
-
-    async def subscribe(self, store_id: bytes32, mode: DownloadMode, ip: str, port: uint16) -> None:
-        subscription = Subscription(store_id, mode, ip, port)
+                self.log.warning(f"Exception while downloading files for {tree_id}: {e} {traceback.format_exc()}.")
+
+    async def upload_files(self, tree_id: bytes32) -> None:
+        singleton_record: Optional[SingletonRecord] = await self.wallet_rpc.dl_latest_singleton(tree_id, True)
+        if singleton_record is None:
+            return
+        root = await self.data_store.get_tree_root(tree_id=tree_id)
+        publish_generation = min(int(singleton_record.generation), 0 if root is None else root.generation)
```

Comment:
yes, since `batch_update` makes now 1 generation to be 1 batch, not 1 single operation.

Reply:
I thought the wallet generations progressed for non-update spends made by other people to get the singleton announcements.

Reply:
I see - so I suppose the root hash remains the same. in this case we'll have empty delta files generated in this step (and getting the full tree pasted in full tree files). downloading part will simply reinsert the same root hash, but with the incremented generation. if the root hash changes after the non-update operations, this will be problematic.